Inspector direct editing

murraystevenson opened the pull request:
This adds support for "direct" editing to the private Inspector API with new `edit()` and `canEdit()` methods on `Inspector::Result` that allow creating an edit directly from a value and checking whether an edit could be made from a value. 

This PR includes two usages of this API, one which allows us to simplify the toggling of bool values in `_InspectorColumn` (and remove `SetMembershipInspector::editSetMembership()` in the process). The other brings drag and drop editing to the AttributeEditor, LightEditor and RenderPassEditor.

![editorDragAndDrop](https://github.com/user-attachments/assets/b1c8349f-c083-4ebf-a0a6-4606e48c0ff3)

We also plan to use this API to provide copy/paste editing in the same editors in a future PR.

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -764,6 +768,8 @@ def __dragBegin( self, widget, event ) :
 			return IECore.StringVectorData( selection[0].paths() )
 
 		index = self.__indexAt( event.line.p0 )
+		self.__dragBeginIndex = index
```
Should we only be setting this in the conditional just before `return value`? When we `return None` we haven't initiated a drag.

murraystevenson replied:
Updated in ea1783d7256b60b9d2c53e41f9ae72856cd6b003.

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -776,6 +782,74 @@ def __dragBegin( self, widget, event ) :
+
+	def __dragEnter( self, widget, event ) :
+
+		self.__currentDragIndex = self.__indexAt( event.line.p0 )
```
I wondered what effect all this new drag handling would have on PathListingWidgets with custom handling, like the ones in the UIEditor (for reordering plugs and presets) and the Catalogue. In testing that, I encountered this crash :

```
#0  0x00007ffff776a0dd in __memchr_evex () from /lib64/libc.so.6
#1  0x00007fffe6d4cd31 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::find(char, unsigned long) const () from /lib64/libstdc++.so.6
#2  0x00007fffe4006ec6 in Gaffer::Path::checkName(IECore::InternedString const&) const () from /home/john/dev/build/gaffer-1.5/lib/libGaffer.so
#3  0x00007fffe4007552 in Gaffer::Path::append(IECore::InternedString const&) () from /home/john/dev/build/gaffer-1.5/lib/libGaffer.so
#4  0x00007fffe3574f8c in (anonymous namespace)::pathForIndex(unsigned long, unsigned long) () from /home/john/dev/build/gaffer-1.5/python/GafferUI/_GafferUI.so
```

I suspect that is because we are storing `__currentDragIndex` and `__dragBeginIndex` as QModelIndex which can be invalidated at any time, when we should be storing QPersistentModelIndex which we can query for validity before use. This same bug caused crashes in our selection handling code in the past - see https://github.com/GafferHQ/gaffer/commit/df7ca8fd99c5d3e70135a30495706d033648487b.

murraystevenson replied:
Thanks, I've updated to storing QPersistentModelIndex in 19d4d539387d8686e6a29eed3cf03bd2c69ee6ef and as part of the overall `__currentDragIndex` rejig in a62e4ee8d2f8152426e9a44c8c5fcfc25f516c75.

I haven't been able to repro the crash, but hopefully you are no longer able to either. :)

johnhaddon replied:
Thanks! I retested my repro with the old branch and it crashed first time. Zero crashes with the new version after a bunch of trying.

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -961,6 +1039,13 @@ def __init__( self ) :
+
+	def setHighlightedIndex( self, index ) :
+
+		self.__highlightedIndex = index
```
I think this is another spot when we can't use a bare QModelIndex, but must instead use QPersistentModelndex.

murraystevenson replied:
Thanks, updated in 0a8d768b78ac41a6bc7c346b76e4990bb767724c.

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -776,6 +782,74 @@ def __dragBegin( self, widget, event ) :
+		if index is None or index == self.__dragBeginIndex :
+			return False
+
+		return self.getColumns()[index.column()].dropSignal()(
```
As far as I can see, we end up calling `drop` on the column under the mouse here, even if that column would have rejected `dragEnter`. I think maybe we need `__currentDragIndex` to track the last column to accept the drag (and be None when the current column rejects the drag) and then deliver drop to that index here?

johnhaddon replied:
Likewise I think we're calling `dragLeaveSignal()` on columns that didn't accept `dragEnterSignal()`.

murraystevenson replied:
Reworked this so `__currentDragIndex` only tracks the last column to accept the drag in a62e4ee8d2f8152426e9a44c8c5fcfc25f516c75.

johnhaddon commented on python/GafferSceneUI/_InspectorColumn.py:
```diff
@@ -420,11 +420,85 @@ def __keyPress( column, pathListing, event ) :
 
+def __dragEnter( path, pathListing, event ) :
+
+	column = pathListing.columnAt( event.line.p0 )
```
Having to re-find the column like this seems awkward. `KeySignal` passes the column as the first argument, so perhaps we could do the same for the drag signals? That would be consistent with most other Widget and Gadget signals (but not the ButtonSignals on PathColumn, which should be brought into line at some point).

murraystevenson replied:
Good idea! Adjusted the drag signals to also pass the column in 5a4b8a66bb4311a801d54de2282c77519c0176dd, bac7e5ff295f649e54e3181cb6f443a1b7fb96a8, and 17e048bc0367cbdde25c6d8fb5da0d5ccee0c9b6.

johnhaddon commented on python/GafferSceneUI/_InspectorColumn.py:
```diff
@@ -420,11 +420,85 @@ def __keyPress( column, pathListing, event ) :
+def __dragEnter( path, pathListing, event ) :
+
+	column = pathListing.columnAt( event.line.p0 )
+	if path is None or not isinstance( column, GafferSceneUI.Private.InspectorColumn ) :
```
We should be guaranteed to have an InspectorColumn shouldn't we, since we only connect to signals on InspectorColumns?

murraystevenson replied:
Yep, this check is redundant. Removed in 71166ddccded22621b83f45bfbd3d27e2a72fe56.

johnhaddon commented on python/GafferSceneUI/_InspectorColumn.py:
```diff
@@ -420,11 +420,85 @@ def __keyPress( column, pathListing, event ) :
+	if inspectionContext is None :
+		return False
+
+	return True
```
Do we actually need to implement `dragMove`? Once we've accepted a drag for a cell, I don't think we care about the precise position do we?

I wasn't going to bring this up, but I also wonder if `event.line` should be relative to the cell, since that's the granularity at which we do enter/leave. I don't think it matters right now though. I almost wonder if we should drop `dragMove()` for simplicity, and because I don't think there's anything we could use it for right now.

johnhaddon replied:
Scratch all that - I see we do use `move` in a later commit, so we can react to keypresses.

johnhaddon reviewed (commented):
Thanks Murray - this is another nice step forwards for the editors. We already chatted a bit about possible refactoring of the `Inspector::Result` methods so they're a bit more uniform - I've made a few comments in that regard, but I don't know if now is the time to address them. Perhaps you'd find it easier to get this over the line and then revisit? Since it's private I think that would be OK if it helps. Other comments inline as usual...
Cheers...
John

murraystevenson commented on include/GafferSceneUI/Private/Inspector.h:
```diff
@@ -340,8 +352,9 @@ class GAFFERSCENEUI_API Inspector::Result : public IECore::RefCounted
 		/// and `false` otherwise.
 		bool editable() const;
```
Agreed, it would be much clearer to rename these, though at the same time I think we'll need to adjust the semantics of the new `canAcquireEdit()` so it just returns whether an edit can be acquired. `editable()` currently also checks to see of the descendants of the edit are editable...

I'll put this on my list for a follow-up PR.

murraystevenson commented on include/GafferSceneUI/Private/Inspector.h:
```diff
@@ -361,6 +374,14 @@ class GAFFERSCENEUI_API Inspector::Result : public IECore::RefCounted
 
+		/// Returns whether a direct edit can be made with the
+		/// specified value.
+		bool canEdit( const IECore::Object *value, std::string &failureReason ) const;
```
Cool, once we rename `editable()` to `canAcquireEdit()` we could rename `nonEditableReason()` to something along the lines of `nonAcquirableReason()` and tackle this.

murraystevenson commented on src/GafferSceneUI/Inspector.cpp:
```diff
@@ -156,6 +156,88 @@ std::string nonEditableReason( const ValuePlug *plug )
+		return false;
+	}
+
+	ConstValuePlugPtr valuePlug;
```
Thanks, adjsted in 0e82abef85842ee81f691e3392ea4bfd95f3aaab.

murraystevenson commented on src/GafferSceneUI/SetMembershipInspector.cpp:
```diff
@@ -127,6 +127,39 @@ HistoryCache g_historyCache(
+	}
+	else
+	{
+		// If we've received a plug without a node, it's expected
```
Yep, for now I've clarified this a bit more with a comment in e195bae91591155825d161ffa8cf91f3b8be6ecf.

murraystevenson commented on src/GafferSceneUI/Inspector.cpp:
```diff
@@ -796,3 +898,47 @@ std::string Inspector::Result::editWarning() const
+
+		if( !plug )
+		{
+			if( const IECore::Data *currentData = runTimeCast<const IECore::Data>( Result::value() ) )
+			{
+				// If we can't acquire an existing edit, create a temporary plug based on the current
+				// value to avoid creating an edit.
+				plug = PlugAlgo::createPlugFromData( "value", Plug::In, Plug::Default, currentData );
+			}
```
Yeah, I wasn't super happy with this as an approach. Your suggestion does sound like a better plan for the future. We'd also want to update `DisableEditFunction` to do the same, and figure out how to deal with edits acquired via returning the `source` plug vs edits acquired from an edit scope.

murraystevenson commented on include/GafferUI/PathColumn.h:
```diff
@@ -172,6 +174,13 @@ class GAFFERUI_API PathColumn : public IECore::RefCounted, public Gaffer::Signal
+		DragDropSignal &dragMoveSignal();
+		DragDropSignal &dragLeaveSignal();
+		DragDropSignal &dropSignal();
+		DragDropSignal &dragEndSignal();
```
Thanks, I've removed `dragEndSignal()` in 19168ff74d9bb16662e6282efeeeeeee643ded7b.

murraystevenson commented:
Thanks for the input, I've addressed the drag and drop related comments inline but will punt the Inspector API related suggestions to a later PR as it feels like some broader reshuffling is necessary now that we have the required functionality in place...

I've added one additional commit on the end which ensures edits on TweakPlugs are made in `Create` mode. This seems like the most consistent behaviour for an interaction like drag and drop or copy / paste. cf766b8abb16dc54b330aabde79cbf31c578f94f

johnhaddon commented:
Thanks Murray, LGTM! Squash and merge away...
